### PR TITLE
fix: validate worker request fields and wipe a refused upload chunk

### LINKED ISSUE #1154: web: validate the worker's read and write request fields instead of coercing them
`buildCommand` now rejects a wrong-typed command field instead of letting wasm-bindgen coerce it. The sibling requests on the same untrusted boundary still coerce.

`EngineHost.beginWrite`, `snapshot`, `download` and `openContentStream` (`packages/client/src/worker/engineHost.ts`) read their arguments straight off the worker message, exactly as `command` did:

```ts
return this.handle.beginWrite(
  this.wasm.NodeId.fromBytes(target.parent),
  target.name,   // unchecked — the coerced-newName bug, unfixed
  undefined,
  size           // unchecked
);
```

`serve.ts` reads `request.target`, `request.size`, `request.folder`, `request.node`, `request.offset` and `request.length` from the message event, and `leaderRelay` forwards a follower's `read`/`write` payload after only a shape test — field validation is deliberately the worker's to do.

The byte arguments are the sharper case. wasm-bindgen marshals `Vec<u8>` through `passArray8ToWasm0`, which is `Uint8Array.prototype.set(arrayLike)`: it coerces rather than throws, so a 16-character **string** reaches `NodeId::from_bytes` as sixteen zero bytes and passes the Rust length check as `NodeId([0u8; 16])` — a request against the wrong node rather than a refusal.

Reachable from a version-skewed follower build with no hostile intent. No key or plaintext exposure, and the engine's own authorization still applies; the defect is that a wrong-typed argument silently becomes a different valid request.

## Scope

- Export the `bytes`/`text` checkers the command codec gained and apply them to every field `EngineHost` reads: `WriteTarget.parent`/`name`/`node`, `beginWrite` size, `snapshot` folder, `download`/`openContentStream` node, `readStream` offset and length.
- Numbers need a finite-integer check, not just `typeof`, since the WASM ABI ToInt32-coerces a string or a `NaN`.
- Cover each with a test asserting refusal, alongside the existing `commandCodec` cases.

Depends on the codec checkers landing first.

### LINKED ISSUE #1155: web: wipe an upload chunk the sender refuses before it is transferred
The engine worker now wipes the upload chunk it takes ownership of, and the leader relay wipes one it accepted but cannot pass on. The **sending** side still drops a chunk it refuses before any transfer, leaving readable plaintext for the collector.

`EngineClient.pushChunk` (`packages/client/src/engineClient.ts`) refuses a stale handle after a leadership swap:

```ts
pushChunk(handle: WriteHandle, chunk: ArrayBuffer): Promise<void> {
  const inner = this.writes.resolve(handle);
  if (inner === undefined) return Promise.reject(unknownHandle('write'));
  return this.current.pushChunk(inner, chunk);
}
```

The chunk is never transferred, so it is never detached and never wiped. `CorrelatedTransport.request` (`packages/client/src/correlatedTransport.ts`) has the same shape on every rejection that fires before `send` runs — a terminal error, a rejected readiness gate, or `send` itself throwing.

The codebase already scrubs on both of the neighbouring paths: `EngineClient.start` wipes the secret buffer it declined to forward, and `LeaderRelay` calls `wipeCarried` on its own unknown-handle refusal. Only the sending side is missing it, and the existing coverage shows the gap — `engineClient.test.ts` asserts the stale-handle rejection without asserting a wipe, while `broadcastTransport.test.ts` asserts the wipe on the relay side only.

Impact is low but real: a failed-over or closing tab leaves up to one chunk of file plaintext in its heap indefinitely, reachable by any later same-origin heap disclosure. AGENTS.md rule 7 — clear sensitive material after use, at the terminal owner. A refusing sender that never hands the buffer on **is** the terminal owner.

## Scope

- Wipe the chunk on `EngineClient.pushChunk`'s unknown-handle refusal, mirroring how `start` handles the secret it declines to forward.
- Wipe on `CorrelatedTransport.request`'s pre-send rejection paths, for the transfer-carrying calls only.
- Cover both with tests asserting the buffer is zero after the rejection, not merely that it rejected.

### PR DESCRIPTION
**Pull request by FSM1**
Two `packages/client` boundary fixes: the worker validates the request fields it reads instead of letting wasm-bindgen coerce them, and the sending side scrubs an upload chunk it refuses before any transfer.

## Validate the worker's read and write request fields

`EngineHost.beginWrite`, `snapshot`, `download`, `openContentStream` and `readStream` read their arguments straight off the worker message, exactly as `command` did before its codec gained checkers. wasm-bindgen coerces rather than throws, so a 16-character string reached `NodeId::from_bytes` as sixteen zero bytes, and a string or `NaN` ToInt32'd into a byte count — a malformed request became a *different valid* request rather than a refusal.

- The command codec's field checkers are now exported and applied to every field the host reads: `WriteTarget.parent`/`name`/`node`, `beginWrite` size, `snapshot` folder, `download`/`openContentStream` node, `readStream` offset and length. Byte fields go through `nodeId`, which is the `bytes` checker composed with `NodeId.fromBytes` — the same check, spelled once.
- Counts get `count`: a non-negative safe integer, since `typeof` alone still admits `NaN`, `1.5` and `-1`.
- `record` refuses a `WriteTarget` that is not an object, so the variant discriminator cannot throw a bare `TypeError`.
- `buffer` guards the two transferred payloads (`start`'s secret, `pushChunk`'s chunk), which went straight into `new Uint8Array(payload)` — a coercion with a sharper edge than the rest, since a `Uint8Array` there is *copied*, so the scrub afterwards zeroes the copy and leaves the sender's plaintext.
- `beginWrite`/`openContentStream` became `async`, so a refused field rejects rather than throwing synchronously — the rest of the surface already did.
- The diagnostic reads `invalid request field …` rather than `invalid command field …`: the same checkers now guard requests that carry no command.

## Wipe an upload chunk the sender refuses before it is transferred

A chunk refused before any transfer is never detached, so the refusing frame is its terminal owner and nothing downstream will ever scrub it (AGENTS.md rule 7). A failed-over or closing tab was leaving up to one chunk of file plaintext in its heap indefinitely.

- `EngineClient.pushChunk` scrubs the chunk its unknown-handle refusal declines to forward, mirroring how `start` handles a secret it declines.
- `CorrelatedTransport.request` scrubs what the send would have transferred on every route to a pre-send rejection: an already-latched terminal error, one latched while the readiness gate was awaited, a rejected gate, and a throwing `send`. A transferred buffer reads as empty, so a send that did run leaves the scrub a no-op — and the happy path is asserted to keep its bytes, since wiping a buffer the receiver is about to seal would corrupt the upload.

- `EngineClient.start` refused a closed client *before* scrubbing the secret it had already declined to forward — the same refuse-before-transfer shape, found by the security pass.
- `EngineTransport` now states outright that a buffer it takes is consumed on every outcome, transferred away or scrubbed in place. A retryable rejection (`leader changed; retry`) hands back a zeroed buffer, so a retry must re-read its source; both upload callers already do, and the contract was the only thing left implicit.

Wiring the two transports to declare what they transfer is 3 lines outside the issues' stated files (`transport.ts` `start`/`command`/`pushChunk`, `broadcastTransport.ts` `overPort`); without it the base-class scrub would never see a real chunk.

## Verification

Each behaviour was broken deliberately and the failing test recorded, then restored — 20 mutants, 20 caught, including an over-wipe mutant that scrubs after a successful send and two that undo the transport wiring. The wipes are asserted on buffer contents after the rejection, never on a spy.

Closes #1154
Closes #1155

> [!NOTE]
> ### Validate worker request fields and zero-fill refused upload chunks
> - Adds upfront type and range validation to `EngineHost` methods (`start`, `beginWrite`, `pushChunk`, `commitWrite`, `abortWrite`, `readStream`, `closeStream`, `snapshot`, `download`, `openContentStream`) using new helpers (`buffer`, `count`, `minted`, `record`, `nodeId`) in [commandCodec.ts](https://github.com/FSM1/cipher-box/pull/1241/files#diff-a2427e8ca2b1c7e229ffba009773c8fe15f5821745326cf1413e23275371319a).
> - Extends `CorrelatedTransport.request` to accept a transfer list and zero-fill any `ArrayBuffer` entries on pre-send failure paths (terminal error, gate rejection, or send exception); successful transfers are unaffected.
> - `EngineClient.pushChunk` now scrubs the chunk buffer when the write handle is unknown or stale; `EngineClient.start` scrubs the secret when called on a closed or non-leader client.
> - Error messages for malformed fields are normalized to `'invalid request field'` across the worker codec.
> - Behavioral Change: callers must not retry with the same buffer after a refused call — buffers are consumed (zeroed or transferred) on every outcome.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized b718e2f.</sup>
> 
>

## Summary by CodeRabbit

* **Security**
  * Sensitive secrets and upload data are now cleared when operations are rejected, fail, or occur after shutdown.
  * Transferred buffers are scrubbed before transmission when dispatch cannot proceed.

* **Bug Fixes**
  * Improved handling of client and transport operations after teardown or closure.
  * Invalid requests are rejected earlier with clearer validation errors.

* **Validation**
  * Added stricter checks for buffers, text, records, identifiers, sizes, and numeric ranges before processing.

### CODE REVIEW

File: packages/client/src/correlatedTransport.ts [hunk 1]
```diff
@@ -72,6 +72,19 @@ export function unknownHandle(kind: HandleKind): EngineRequestError {
...
+ * Scrubs the buffers a send would have transferred. A request that rejects
+ * before its send leaves this frame their terminal owner — nothing detaches
+ * them and no callee can reach them — so the plaintext is cleared here
+ * (AGENTS.md 7). A transferred buffer reads as empty, so a send that did run
+ * leaves this a no-op.
+ */
+function wipeTransfer(transfer: Transferable[] | undefined): void {
+  for (const item of transfer ?? []) {
+    if (item instanceof ArrayBuffer && item.byteLength > 0) new Uint8Array(item).fill(0);
+  }
```
[the remark this answers is not in the document]
Comment (FSM1) [reply packages/client/src/correlatedTransport.ts:102]: Verified and taken in b718e2f. Reproduced the realm behaviour here rather than taking it on the script's word:

```
instanceof:              false
byteLength getter brand: 4
brand on {} / null / MessagePort: null null null
bytes after same-realm fill: [0, 0, 0, 0]
```

So all three legs hold: `instanceof` does skip a foreign buffer, the getter brands one correctly across realms, and `new Uint8Array(foreign).fill(0)` does zero its bytes — the scrub still works once the branch is entered.

`wipeTransfer` now brands by the `byteLength` getter, which returns `null` for every non-buffer transferable (a `MessagePort` included) rather than throwing:

```ts
const byteLengthOf = Object.getOwnPropertyDescriptor(ArrayBuffer.prototype, 'byteLength')?.get;
```

I did not use `Object.prototype.toString`, the other realm-independent option: `Symbol.toStringTag` is writable, so a buffer could be made to answer something else and skip the scrub — the wrong direction to fail on a path whose whole job is clearing plaintext.

Regression coverage added: `wipes a chunk minted in another realm, which instanceof does not answer for` builds a real foreign buffer through `node:vm`, asserts `instanceof` is false for it, drives a refused `pushChunk`, and asserts the bytes are zeroed. Verified as a true negative control — restoring the `instanceof` branch fails it.

Gates on b718e2f: `pnpm lint` 0 · `pnpm typecheck` 0 · `pnpm --filter @cipherbox/client test` 0 (29 files, **441 tests**).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/client/src/worker/commandCodec.ts [hunk 2]
```diff
@@ -28,32 +28,61 @@ import type {
...
  */
 function invalidField(field: string, value: unknown): Error {
-  return new Error(`invalid command field ${field}: ${value === null ? 'null' : typeof value}`);
+  return new Error(`invalid request field ${field}: ${value === null ? 'null' : typeof value}`);
+}
+
+/** An untrusted wire object; a non-object carries no fields at all. */
+export function record(value: unknown, field: string): Record<string, unknown> {
+  if (typeof value !== 'object' || value === null) throw invalidField(field, value);
+  return value as Record<string, unknown>;
```
[the remark this answers is not in the document]
Comment (FSM1) [reply packages/client/src/worker/commandCodec.ts:45]: Verified and taken in b718e2f — and it was an inconsistency with this file's own stated rule.

The module header says every field the worker reads off a request passes through a checker; `beginWrite` does exactly that (`record(target, 'target')`, `count(size, 'size')`). The command envelope was the exception, and it arrives just as untrusted: `serve.ts:53` hands `request.command` straight to `host.command`, which passes it to `buildCommand` unvalidated.

`buildCommand` now validates before the switch:

```ts
text(record(descriptor, 'command').kind, 'command.kind');
```

Confirmed the failure modes you named. Before: `null` gave `TypeError: Cannot read properties of null`, and a primitive or a missing `kind` fell through to `unknown command kind: undefined` — neither naming the field. After, all of them answer `invalid request field …` like every other malformed input.

Regression coverage added, in the file's existing `refuses` idiom, which builds against a permissive fake WASM so a factory firing would be visible:

```
refuses(null)            → invalid request field command: null
refuses(42)              → invalid request field command: number
refuses('rename')        → invalid request field command: string
refuses({ kind: 12345 }) → invalid request field command.kind: number
refuses({})              → invalid request field command.kind: undefined
```

Verified as a true negative control — removing the guard line fails the case.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/client/src/worker/engineHost.ts [hunk 3]
```diff
@@ -99,33 +108,35 @@ export class EngineHost implements EngineHostLike {
...
       undefined,
-      size
+      reserved
     );
   }
 
-  pushChunk(handle: WriteHandle, chunk: ArrayBuffer): Promise<void> {
-    return this.scrubbing(chunk, (view) => this.handle.pushChunk(handle, view));
+  async pushChunk(handle: WriteHandle, chunk: ArrayBuffer): Promise<void> {
+    return this.scrubbing(buffer(chunk, 'chunk'), (view) => this.handle.pushChunk(handle, view));
```
[the remark this answers is not in the document]
Comment (FSM1) [reply packages/client/src/worker/engineHost.ts]: Verified and taken in b718e2f, across all five sites — and it turned up a sixth thing worth fixing.

Handles were the one field class crossing to WASM unchecked while their neighbours were validated: `pushChunk` checked `chunk` but not `handle`, `readStream` checked `offset` and `length` but not `handle`, and `commitWrite` / `abortWrite` / `closeStream` checked nothing at all.

A shared validator, sitting with the other checkers and reusing the wording:

```ts
export function minted(value: unknown, field: string): bigint {
  if (typeof value !== 'bigint') throw invalidField(field, value);
  return value;
}
```

It replaces the private `opId`, which was the same check under a narrower name, so there is one bigint checker rather than two.

**The sixth thing:** `commitWrite` was the only one of the five that was not `async`, so its refusal threw **synchronously out of the call** while the other four rejected. A caller doing `await host.commitWrite(x)` inside a `try` still catches it, but a caller holding the promise does not — and a boundary should refuse in one shape. It is `async` now. This surfaced because the new test asserted `rejects` uniformly and `commitWrite` alone failed it; worth noting since it was not visible from reading.

Behaviour tests added in the file's `it.each` idiom, one per operation, each asserting the refusal **and** `expect(calls).toEqual([])` — the permissive fake records every WASM call, so an empty list is the evidence that none ran:

```
refuses a pushChunk    carrying a handle the engine never minted
refuses a commitWrite  …
refuses an abortWrite  …
refuses a readStream   …
refuses a closeStream  …
```

Verified as true negative controls — reverting `minted` at the five sites fails all five.

Gates on b718e2f: `pnpm lint` 0 · `pnpm typecheck` 0 · `pnpm --filter @cipherbox/client test` 0 (29 files, **441 tests**) · `pnpm lint:tracker-refs` 0.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by FSM1** [comment]
## Review disposition — CodeRabbit on `816668c`

3 actionable, all **Security & Privacy / Major**, **no nitpicks and no outside-diff notes**. All three verified and taken in b718e2f. This was a well-aimed review: each one was a field reaching past the boundary this PR exists to guard.

| Finding | Disposition |
| --- | --- |
| `wipeTransfer` brands with `instanceof ArrayBuffer`, false across realms | **Taken.** Reproduced the realm behaviour locally rather than trusting the script: `instanceof` false, getter brand `4`, `null` for `{}`/`null`/`MessagePort`, and the fill does zero foreign bytes. Now branded by the `byteLength` getter. Chose it over `Object.prototype.toString` because `Symbol.toStringTag` is writable — a buffer could be made to answer something else and skip the scrub, the wrong way to fail on a plaintext-clearing path. |
| `buildCommand` reads `kind` before validating the envelope | **Taken.** An inconsistency with this file's own header, which says every field the worker reads passes a checker — `beginWrite` does, the envelope did not, and `serve.ts:53` hands it straight from the wire. `null` gave a `TypeError`; a primitive or missing `kind` gave `unknown command kind: undefined`. Both answer `invalid request field …` now. |
| write and stream handles forwarded to WASM unchecked | **Taken at all five sites.** `pushChunk` validated `chunk` but not `handle`; `readStream` validated `offset`/`length` but not `handle`; `commitWrite`, `abortWrite`, `closeStream` validated nothing. One shared `minted()` checker now, replacing the private `opId` so there is one bigint check rather than two. |

### One thing the review did not name, found while fixing it

`commitWrite` was the only one of the five **not `async`**, so its refusal threw *synchronously out of the call* while the other four rejected. A boundary should refuse in one shape; it is `async` now. This surfaced only because the new test asserted `rejects` uniformly and `commitWrite` alone failed — it was not visible from reading the diff.

### Coverage added

7 cases, each verified as a true negative control (reverting the guard fails it, restoring it passes):

- 5 × `refuses a {pushChunk,commitWrite,abortWrite,readStream,closeStream} carrying a handle the engine never minted` — each also asserts `calls` is empty, so an empty WASM call log is the evidence none ran
- `refuses an envelope that is not a command before it reads a kind off it` — `null`, a number, a string, a non-string `kind`, and a missing `kind`
- `wipes a chunk minted in another realm, which instanceof does not answer for` — a real foreign buffer via `node:vm`, asserting `instanceof` is false for it and that the bytes end zeroed

### Gates on `b718e2f`

`pnpm lint` 0 · `pnpm typecheck` 0 · `pnpm --filter @cipherbox/client test` 0 (29 files, **441 tests**) · `pnpm lint:tracker-refs` 0.

---
_Generated by [Claude Code](https://claude.ai/code)_
